lib/maths/CBoostedTreeFactory.cc

```diff
@@ -708,6 +709,52 @@ void CBoostedTreeFactory::initializeUnsetDownsampleFactor(core::CDataFrame& fram
     }
 }
 
+void CBoostedTreeFactory::initializeUnsetEta(core::CDataFrame& frame) {
+
+    if (m_TreeImpl->m_EtaOverride == boost::none) {
+        double searchIntervalSize{5.0 * MAX_ETA_SCALE / MIN_ETA_SCALE};
+        double logMaxEta{std::log(std::sqrt(searchIntervalSize) * m_TreeImpl->m_Eta)};
+        double logMinEta{logMaxEta - std::log(searchIntervalSize)};
+        double meanLogEta{(logMaxEta + logMinEta) / 2.0};
+        double mainLoopSearchInterval{std::log(0.2 * searchIntervalSize)};
+        LOG_TRACE(<< "mean log eta = " << meanLogEta);
+
+        auto applyEta = [](CBoostedTreeImpl& tree, double eta) {
+            tree.m_Eta = std::exp(eta);
+            tree.m_EtaGrowthRatePerTree = 1.0 + tree.m_Eta / 2.0;
+            tree.m_MaximumNumberTrees = computeMaximumNumberTrees(tree.m_Eta);
+            return true;
+        };
+
+        double eta{m_TreeImpl->m_Eta};
+
+        TVector fallback;
+        fallback(MIN_REGULARIZER_INDEX) = logMinEta;
+        fallback(BEST_REGULARIZER_INDEX) = meanLogEta;
+        fallback(MAX_REGULARIZER_INDEX) = logMaxEta;
+
+        m_LogEtaSearchInterval =
+            this->testLossLineSearch(frame, applyEta, logMinEta, logMaxEta,
+                                     -mainLoopSearchInterval / 2.0,
+                                     mainLoopSearchInterval / 2.0)
+                .value_or(fallback);
+        m_LogEtaSearchInterval = min(m_LogEtaSearchInterval, TVector{0.0});
+        LOG_TRACE(<< "log eta search interval = ["
+                  << m_LogEtaSearchInterval.toDelimited() << "]");
+        m_TreeImpl->m_Eta = std::exp(m_LogEtaSearchInterval(BEST_REGULARIZER_INDEX));
+        m_TreeImpl->m_EtaGrowthRatePerTree = 1.0 + m_TreeImpl->m_Eta / 2.0;
+        m_TreeImpl->m_MaximumNumberTrees = computeMaximumNumberTrees(m_TreeImpl->m_Eta);
```

Comment:
Can you use `applyEta` here?

Reply:
Good suggestion. See [fa55c62](https://github.com/elastic/ml-cpp/pull/948/commits/fa55c62f1d40558d8eebe87151f72c34888ecb1f).